cdk/l3-cdk-resources/atlas-basic/src/index.ts

```diff
@@ -1,98 +1,414 @@
       {
-        password: defaults.password,
-        ...props.dbUserProps,
+          apiKeys: props.apiKeys,
+          databaseName: props.dbUserProps?.databaseName || defaults.dbName,
+          projectId: mProject.ref,
+          username: props.dbUserProps?.username || defaults.username,
+          roles: props.dbUserProps?.roles || defaults.roles,
+          password: props.dbUserProps?.password || defaults.password,
+          ...props.dbUserProps,
       });
     // Create a new MongoDB Atlas Project IP Access List
     new ipAccessList.CfnProjectIpAccessList(this, 'ip-access-list-'.concat(id),
-      {
-        ...props.ipAccessListProps,
-      },
+        {
+            apiKeys: props.apiKeys,
+            projectId: mProject.ref,
+            accessList: props.ipAccessListProps?.accessList || defaults.accessList,
+            ...props.ipAccessListProps,
+        }
     );
 
-    //
-    // new cloudformation.CfnTypeActivation(this, 'ProjectActivation', {
-    //     type: mProject.cfnResourceType,
-    //     executionRoleArn: ""
-    // });
-    //
-    // new cloudformation.CfnTypeActivation(this, 'ClusterActivation', {
-    //     type: mCluster.cfnResourceType,
-    //     executionRoleArn: ""
-    // });
-    //
-    // new cloudformation.CfnTypeActivation(this, 'DatabaseUserActivation', {
-    //     type: dbUser.cfnResourceType,
-    //     executionRoleArn: "",
-    // });
-    //
-    // new cloudformation.CfnTypeActivation(this, 'ProjectIpAccessListActivation', {
-    //     type: ipAccess.cfnResourceType,
-    //     executionRoleArn: "",
-    // });
-
   }
-}
-
-// export interface ProjectProps extends Omit<project.CfnProjectProps, 'name'> {
-//     readonly  name? : string
-// }
-//
-// export interface ClusterProps extends Omit<atlas.CfnClusterProps, 'projectId' | 'name' | 'apiKeys'> {
-//     readonly name? : string
-//     readonly projectId? : string
-//     readonly apiKeys?: atlas.ApiKeyDefinition;
-// }
-//
-// export interface DatabaseUserProps extends Omit<user.CfnDatabaseUserProps, 'databaseName' | 'roles' | 'username' | 'projectId'> {
-//     readonly databaseName? : string
-//     readonly roles? : user.RoleDefinition[]
-//     readonly username? : string
-//     readonly projectId? : string
-// }
-//
-// export interface IpAccessListProps extends Omit<ipAccessList.CfnProjectIpAccessListProps, 'accessList' | 'apiKeys' | 'projectId'> {
-//     readonly accessList?: ipAccessList.AccessListDefinition[];
-//     readonly apiKeys?: ipAccessList.ApiKeyDefinition;
-//     readonly projectId?: string;
-// }
-
-export interface AtlasBasicProps extends cdk.StackProps{
-  readonly projectId : string;
-  readonly apiKeys : atlas.ApiKeyDefinition;
-  // readonly projectProps: project.CfnProjectProps;
-  readonly clusterProps: atlas.CfnClusterProps;
-  readonly dbUserProps: user.CfnDatabaseUserProps;
-  readonly ipAccessListProps: ipAccessList.CfnProjectIpAccessListProps;
```

Comment:
I am curious as to why we couldn't reuse/extend the L1s with additional properties(if any) and instead re-wrote the interfaces? For example, with ApiKeyDefinition.

Reply:
For instance, projectId is a required field for Cluster. If I were to use the CfnClusterProps, that would force me to inject projectId and it would make no sense to create another project in this app.

Reply:
Hi @SuperMohit, I am not sure I am following as I don't understand your point: you can create a new project, read the project and use the projectId to create the cluster, right?

Reply:
@andreaangiolillo That's what I am doing right now. But If I reuse CfnClusterProps from L1 in the constructor, which has `projectId` required field. I am forced to add projectId from outside the Library. I tried to make projectId optional using <Omit> but it seems jsii doesn't allow it. So, I was forced to fall back to DTO pattern.